In [66]:
import numpy as np
import open3d as o3d
import os
import json
from scipy.spatial import cKDTree

def compute_metrics(gt_pcd, gen_pcd):
    """
    Computes various metrics to evaluate the accuracy of a generated point cloud against the ground truth.
    Handles cases where the point counts differ.
    
    Args:
        gt_points (numpy.ndarray): Ground truth points, shape (N1, 3).
        gen_points (numpy.ndarray): Generated points, shape (N2, 3).
        
    Returns:
        dict: A dictionary containing the computed metrics.
    """
    # Convert to numpy arrays
    gt_points = np.asarray(gt_pcd.points)
    gen_points = np.asarray(gen_pcd.points)

    # Compute pairwise distances using k-D tree for efficiency
    gt_kdtree = cKDTree(gt_points)
    gen_kdtree = cKDTree(gen_points)


    gt_set = set(map(tuple, gt_points))
    gen_set = set(map(tuple, gen_points))
    
    # For Chamfer Distance and Hausdorff Distance
    gen_to_gt_distances, _ = gen_kdtree.query(gt_points, k=1)
    gt_to_gen_distances, _ = gt_kdtree.query(gen_points, k=1)

    
    aabb = gt_pcd.get_axis_aligned_bounding_box()
    aabb_extent = aabb.get_extent()  # [dx, dy, dz]
    aabb_diagonal = np.linalg.norm(aabb_extent)  # Diagonal length
    
    # Metrics
    # Chamfer Distance (First Term)
    chamfer_distance = np.mean(gen_to_gt_distances ** 2) + np.mean(gt_to_gen_distances ** 2)
    
    # Hausdorff Distance
    hausdorff_distance = max(np.max(gen_to_gt_distances), np.max(gt_to_gen_distances))
    hausdorff_bb_ratio = hausdorff_distance / aabb_diagonal

    # MSE and RMSE
    mse = np.mean(gen_to_gt_distances ** 2)  # Only for nearest points
    rmse = np.sqrt(mse)
    
    # MAE
    mae = np.mean(gen_to_gt_distances)  # Absolute distance
    
    # Signal-to-Noise Ratio (SNR)
    signal_power = np.sum(np.linalg.norm(gt_points, axis=1) ** 2)
    noise_power = np.sum(gen_to_gt_distances ** 2)
    snr = 10 * np.log10(signal_power / noise_power) if noise_power > 0 else float('inf')

    # Exact matches
    exact_matches = gt_set & gen_set
    matched_count = len(exact_matches)
    
    # Missing and extra points
    missing_count = len(gt_set - gen_set)
    extra_count = len(gen_set - gt_set)
    
    # Completeness (Recall)
    completeness = matched_count / len(gt_set) if len(gt_set) > 0 else 1.0
    
    # Accuracy (Precision)
    accuracy = matched_count / len(gen_set) if len(gen_set) > 0 else 1.0
    
    # F1 Score
    if completeness + accuracy > 0:
        f1_score = 2 * (completeness * accuracy) / (completeness + accuracy)
    else:
        f1_score = 0.0
    
        
    return {
        "MSE": mse,
        "RMSE": rmse,
        "MAE": mae,
        "Chamfer Distance": chamfer_distance,
        "Hausdorff Distance": hausdorff_distance,
        "Bounding Box Diagonal": aabb_diagonal,
        "Hausdorff / BB Ratio": hausdorff_bb_ratio,
        "SNR": snr,
        "Exact Matches": matched_count,
        "Missing Points": missing_count,
        "Extra Points": extra_count,
        "Completeness/Recall": completeness,
        "Accuracy/Precision": accuracy,
        "F1 Score": f1_score,
        "Chamfer Distance": chamfer_distance
    }


def ply_metrics(gt_path, gen_dir, name, output_file):
    """
    Computes metrics for all .ply files in the given directories matching a specific naming pattern.
    
    Args:
        gt_dir (str): Directory containing the ground truth point clouds.
        gen_dir (str): Directory containing the generated point clouds.
        name (str): Base name for the point cloud files (e.g., "example").
        output_file (str): Path to the file where results will be saved.
        
    Returns:
        None
    """
    # Define the file patterns to process
    file_patterns = [f"{name}_density.ply", f"{name}_final.ply", f"{name}_floorseg.ply"]
    results = {}
    gt_pcd = o3d.io.read_point_cloud(gt_path)

    for file_pattern in file_patterns:
        gen_path = os.path.join(gen_dir, file_pattern)
        print()
        print(gen_path)
        
        # Check if both files exist
        if not os.path.exists(gen_path):
            print(f"Skipping {file_pattern} as gen file is missing.")
            continue
        
        # Load point clouds
        gen_pcd = o3d.io.read_point_cloud(gen_path)
        
        # Convert to numpy arrays
        
        # Compute metrics
        metrics = compute_metrics(gt_pcd, gen_pcd)
        results[file_pattern] = metrics
        print(f"Metrics for {file_pattern}:")
        for metric, value in metrics.items():
            print(f"  {metric}: {value}")
    
    # Save results to a file
    with open(output_file, "w") as f:
        json.dump(results, f, indent=4)
    print(f"Results saved to {output_file}")


In [67]:
bicycle_metrics = ply_metrics("data/points_bicycle_gt.ply", "output/bicycle/cluster", "bicycle", "output/bicycle/cluster/metrics.json")


output/bicycle/cluster\bicycle_density.ply
Metrics for bicycle_density.ply:
  MSE: 0.0
  RMSE: 0.0
  MAE: 0.0
  Chamfer Distance: 0.297468168421502
  Hausdorff Distance: 1.8987165015454919
  Bounding Box Diagonal: 3.8078062307677456
  Hausdorff / BB Ratio: 0.49863789974487877
  SNR: inf
  Exact Matches: 6334
  Missing Points: 0
  Extra Points: 9256
  Completeness/Recall: 1.0
  Accuracy/Precision: 0.4062860808210391
  F1 Score: 0.5778142674694399

output/bicycle/cluster\bicycle_final.ply
Metrics for bicycle_final.ply:
  MSE: 1.4679513616464817e-05
  RMSE: 0.0038313853390731685
  MAE: 0.0002952980516493955
  Chamfer Distance: 0.0035690604638450216
  Hausdorff Distance: 0.430407705370072
  Bounding Box Diagonal: 3.8078062307677456
  Hausdorff / BB Ratio: 0.11303298521135395
  SNR: 50.0303663037498
  Exact Matches: 6281
  Missing Points: 53
  Extra Points: 506
  Completeness/Recall: 0.9916324597410799
  Accuracy/Precision: 0.9254457050243112
  F1 Score: 0.9573965398978738

output/bicycle/

MSE: 1.2573330356184421e-05
RMSE: 0.003545889219389746
MAE: 0.00021460280284468124
Chamfer Distance: 0.004557185225284462
Hausdorff Distance: 0.4778152120945047
Maximum Distance GT: 0.3307482331216485
Maximum Distance Gen: 0.3893983222353729
SNR: 50.70297969988866
Exact Matches: 6308
Missing Points: 26
Extra Points: 689
Completeness: 0.9958951689295864
Accuracy: 0.9015292268114906

In [69]:
bonsai_metrics = ply_metrics("data/points_bonsai_gt.ply", "output/bonsai/cluster", "bonsai", "output/bonsai/cluster/metrics.json")



output/bonsai/cluster\bonsai_density.ply
Metrics for bonsai_density.ply:
  MSE: 0.0
  RMSE: 0.0
  MAE: 0.0
  Chamfer Distance: 0.22602038574583713
  Hausdorff Distance: 1.9056247417840848
  Bounding Box Diagonal: 2.9727479402394965
  Hausdorff / BB Ratio: 0.6410313891700351
  SNR: inf
  Exact Matches: 25220
  Missing Points: 0
  Extra Points: 101650
  Completeness/Recall: 1.0
  Accuracy/Precision: 0.19878615906045558
  F1 Score: 0.3316457360773226

output/bonsai/cluster\bonsai_final.ply
Metrics for bonsai_final.ply:
  MSE: 0.0002942565909399683
  RMSE: 0.017153908911381344
  MAE: 0.0029390689556676817
  Chamfer Distance: 0.00035447929147648313
  Hausdorff Distance: 0.2453098323343594
  Bounding Box Diagonal: 2.9727479402394965
  Hausdorff / BB Ratio: 0.08251955337814354
  SNR: 40.81516682514785
  Exact Matches: 24363
  Missing Points: 857
  Extra Points: 128
  Completeness/Recall: 0.9660190325138779
  Accuracy/Precision: 0.994773590298477
  F1 Score: 0.9801854720283237

output/bonsai/

MSE: 0.0001709805336266227
RMSE: 0.013075952494048864
MAE: 0.001834674022331613
Chamfer Distance: 0.00023835957867605645
Hausdorff Distance: 0.2453098323343594
SNR: 43.172962135626975
Exact Matches: 24531
Missing Points: 689
Extra Points: 135
Completeness: 0.972680412371134
Accuracy: 0.9945268791048407

In [70]:
kitchen_metrics = ply_metrics("data/points_kitchen_gt.ply", "output/kitchen/cluster", "kitchen", "output/kitchen/cluster/metrics.json")



output/kitchen/cluster\kitchen_density.ply
Metrics for kitchen_density.ply:
  MSE: 0.0
  RMSE: 0.0
  MAE: 0.0
  Chamfer Distance: 0.3150447259020694
  Hausdorff Distance: 1.3619621770644579
  Bounding Box Diagonal: 2.7679881761429908
  Hausdorff / BB Ratio: 0.49204046057821765
  SNR: inf
  Exact Matches: 37638
  Missing Points: 0
  Extra Points: 104054
  Completeness/Recall: 1.0
  Accuracy/Precision: 0.2656324986590633
  F1 Score: 0.41976244911615457

output/kitchen/cluster\kitchen_final.ply
Metrics for kitchen_final.ply:
  MSE: 0.0007681347634677821
  RMSE: 0.027715244243336233
  MAE: 0.008878417324621793
  Chamfer Distance: 0.0009621539225524939
  Hausdorff Distance: 0.41988873434785656
  Bounding Box Diagonal: 2.7679881761429908
  Hausdorff / BB Ratio: 0.151694554899777
  SNR: 37.65447349422094
  Exact Matches: 32916
  Missing Points: 4722
  Extra Points: 766
  Completeness/Recall: 0.8745416865933365
  Accuracy/Precision: 0.9772578825485423
  F1 Score: 0.9230510375771173

output/ki

MSE: 0.0010305423646034932
RMSE: 0.03210206168774045
MAE: 0.01045092367960673
Chamfer Distance: 0.0012296781090043564
Hausdorff Distance: 0.4269672716415563
SNR: 36.37818919929256
Exact Matches: 32827
Missing Points: 4811
Extra Points: 767
Completeness: 0.8721770551038843
Accuracy: 0.9771685420015479

In [71]:
garden_metrics = ply_metrics("data/points_garden_gt.ply", "output/garden/cluster", "garden", "output/garden/cluster/metrics.json")



output/garden/cluster\garden_density.ply
Metrics for garden_density.ply:
  MSE: 0.0
  RMSE: 0.0
  MAE: 0.0
  Chamfer Distance: 2.5403588049146486
  Hausdorff Distance: 4.34449096644896
  Bounding Box Diagonal: 4.9021048569431915
  Hausdorff / BB Ratio: 0.8862501095413241
  SNR: inf
  Exact Matches: 14055
  Missing Points: 0
  Extra Points: 51018
  Completeness/Recall: 1.0
  Accuracy/Precision: 0.21598819787008436
  F1 Score: 0.3552471944191689

output/garden/cluster\garden_final.ply
Metrics for garden_final.ply:
  MSE: 0.0012241690711216346
  RMSE: 0.03498812757381616
  MAE: 0.010007045991085348
  Chamfer Distance: 0.0013157157168922852
  Hausdorff Distance: 0.3944644905621165
  Bounding Box Diagonal: 4.9021048569431915
  Hausdorff / BB Ratio: 0.08046839104296373
  SNR: 36.222138912315366
  Exact Matches: 12720
  Missing Points: 1335
  Extra Points: 37
  Completeness/Recall: 0.9050160085378869
  Accuracy/Precision: 0.9970996315748216
  F1 Score: 0.9488288825898852

output/garden/clust

MSE: 0.0011595296083539836
RMSE: 0.034051866444498806
MAE: 0.009683864530969196
Chamfer Distance: 0.0012840190732325805
Hausdorff Distance: 0.5481564818660584
SNR: 36.45773451258344
Exact Matches: 12733
Missing Points: 1322
Extra Points: 38
Completeness: 0.9059409462824618
Accuracy: 0.9970245086524157